src/ServiceControl/Monitoring/HeartbeatMonitor.cs

```diff
@@ -0,0 +1,53 @@
+namespace ServiceControl.Monitoring
+{
+    using System;
+    using System.Threading;
+
+    class HeartbeatMonitor
+    {
+        volatile RecordedHeartbeat heartbeat = new RecordedHeartbeat(HeartbeatStatus.Unknown, null);
+
+        public void MarkAlive(DateTime timestamp)
+        {
+            var newReading = new RecordedHeartbeat(HeartbeatStatus.Alive, timestamp);
+            var done = false;
+
+            while (!done)
+            {
+                var priorReading = heartbeat;
+
+                if (priorReading.Timestamp.GetValueOrDefault() <= timestamp)
+                {
+                    done = Interlocked.CompareExchange(ref heartbeat, newReading, priorReading) == priorReading;
+                }
+                else
+                {
+                    done = true;
+                }
+            }
+        }
+
+        public RecordedHeartbeat Age(DateTime cutoff)
+        {
+            var done = false;
+
+            while (!done)
+            {
+                var priorReading = heartbeat;
+                if (priorReading.Timestamp.GetValueOrDefault() < cutoff)
+                {
+                    var newReading = new RecordedHeartbeat(HeartbeatStatus.Dead, priorReading.Timestamp);
+                    done = Interlocked.CompareExchange(ref heartbeat, newReading, priorReading) == priorReading;
```

Comment:
It's to ensure consistency. If I read the timestamp and see that my timestamp is from further into the future then I can overwrite that record. If, in the time between the read and the write another heartbeat is written in, I can't overwrite that one without checking it. BUT that new timestamp might still be from before the cutoff. 

If we don't do the check then we run the risk of going slightly out of sync when the queue is backed up. We declare the endpoint dead moments after the latest heartbeat is received.